src/main/java/nl/workingtalent/wtacademy/user/UserService.java

```diff
@@ -21,24 +22,35 @@ public Optional<User> findUserById(long id) {
 		return repository.findById(id);
 	}
 
-	public List<User> findUserByFirstName(String name) {
-		return repository.findUserByFirstName(name);
-	}
+	public List<User> searchUser(SearchUserDto searchUserDto) {
+		String firstName = searchUserDto.getFirstName();
+		String lastName = searchUserDto.getLastName();
+		String email = searchUserDto.getEmail();
+		String role = searchUserDto.getRole();
 
-	public List<User> findUserByLastName(String name) {
-		return repository.findUserByLastName(name);
-	}
+		// Constructing the query based on provided criteria
+		Specification<User> spec = Specification.where(null);
 
-	public Optional<User> findUserByEmail(String email) {
-		return repository.findUserByEmail(email);
-	}
+		if (firstName != null && !firstName.isEmpty()) {
+			spec = spec.and((root, query, builder) -> builder.like(root.get("firstName"), "%" + firstName + "%"));
+		}
+
+		if (lastName != null && !lastName.isEmpty()) {
+			spec = spec.and((root, query, builder) -> builder.like(root.get("lastName"), "%" + lastName + "%"));
+		}
+
+		if (email != null && !email.isEmpty()) {
+			spec = spec.and((root, query, builder) -> builder.equal(root.get("email"), email));
+		}
 
-	public List<User> findUserByRole(Role role) {
+		if (role != null && !role.isEmpty()) {
+			spec = spec.and((root, query, builder) -> builder.equal(root.get("role"), "%" + role + "%"));
+		}
```

Comment:
For builder.equal(). Do not use the % around the variable this is used for builder.like to indicate any characters are allowed before or after the string in the variable. Searching for role works if the % are removed.